### PR TITLE
Added span.parentId parameter.  Added descriptions for span parameter…

### LINKED ISSUE #16: ParentId missing from Span definition
Swagger span definition is missing parentId

### PR DESCRIPTION
**Pull request by singram**
…s and link to published API.

Resolved #16

### CODE REVIEW

File: zipkin-api.yaml [hunk 1]
```diff
@@ -248,17 +248,30 @@ definitions:
     properties:
       traceId:
         type: string
+        description: 'Randomly generated globally unique transaction/context id'
       name:
         type: string
+        description: 'RPC name'
+      parentId:
```
Comment (adriancole) [inline zipkin-api.yaml]: is there an optional flag in swagger?

File: zipkin-api.yaml [hunk 2]
```diff
@@ -248,17 +248,30 @@ definitions:
     properties:
       traceId:
         type: string
+        description: 'Randomly generated globally unique transaction/context id'
```
Comment (adriancole) [inline zipkin-api.yaml]: In code we used `Unique 8-byte identifier for a trace, set on all spans within it.` Good point that it should be random.. but it is also hex-encoded.. I wouldn't conflate it for a transaction or context id. Also global is relative. Maybe..

`Randomly generated, unique 8-byte identifier for a trace, set on all spans within it. Encoded into 16 lowercase hex characters`

File: zipkin-api.yaml [hunk 3]
```diff
@@ -248,17 +248,30 @@ definitions:
...
       name:
         type: string
+        description: 'RPC name'
```
Comment (adriancole) [inline zipkin-api.yaml]: In code, we say..

```
  /**
   * Span name in lowercase, rpc method for example.
   *
   * <p>Conventionally, when the span name isn't known, name = "unknown".
   */
```

File: zipkin-api.yaml [hunk 4]
```diff
@@ -248,17 +248,30 @@ definitions:
...
+      parentId:
+        type: string
+        maxLength: 16
+        minLength: 16
+        description: 'Parent span id (16 digit hex value)'
```
Comment (adriancole) [inline zipkin-api.yaml]: The parent's span id (16 digit hex value). Unset if this the root span in a trace.

Comment (adriancole) [inline zipkin-api.yaml]: well, rather we should be consistent with something like `Encoded into 16 lowercase hex characters` remembering hex aren't all digits

File: zipkin-api.yaml [hunk 5]
```diff
@@ -248,17 +248,30 @@ definitions:
...
       id:
         type: string
+        maxLength: 16
+        minLength: 16
+        description: 'Id of current span (16 digit hex value)'
```
Comment (adriancole) [inline zipkin-api.yaml]: In code, we say..

```
  /**
   * Unique 8-byte identifier of this span within a trace.
   *
   * <p>A span is uniquely identified in storage by ({@linkplain #traceId}, {@code #id}).
   */
```

File: zipkin-api.yaml [hunk 6]
```diff
@@ -248,17 +248,30 @@ definitions:
...
       timestamp:
         type: integer
         format: int64
+        description: 'Start time of span as UNIX timestamp in **milliseconds**'
```
Comment (adriancole) [inline zipkin-api.yaml]: this is incorrect as the unit is microseconds.

`Epoch microseconds of the start of this span, possibly absent if this an incomplete span.`

File: zipkin-api.yaml [hunk 7]
```diff
@@ -248,17 +248,30 @@ definitions:
...
       duration:
         type: integer
+        description: 'Duration of span in **milliseconds**'
```
Comment (adriancole) [inline zipkin-api.yaml]: `Measurement in microseconds of the critical path, if known.`

also This field should be int64 vs integer to support spans longer than 35 minutes.

File: zipkin-api.yaml [hunk 8]
```diff
@@ -248,17 +248,30 @@ definitions:
...
       debug:
         type: boolean
+        description: 'Indicator whether this trace was sampled from standard traffic or forced.  Usually coupled to the X-B3-Flags header'
```
Comment (adriancole) [inline zipkin-api.yaml]: might want to merge with the code comment: `True is a request to store this span even if it overrides sampling policy.`

**Comment by adriancole** [comment]
thanks for taking a stab. I made some clarifying comments

**Comment by singram** [comment]
@adriancole Thanks for the thorough feedback, appreciated!  I'll go through, revise and get back to you with any questions on the above.

File: zipkin-api.yaml (hunk 1 above)
Comment (singram) [inline zipkin-api.yaml]: There is for parameter types against endpoints but not for type definitions (e.g. Span) oddly enough.

File: zipkin-api.yaml (hunk 8 above)
Comment (singram) [inline zipkin-api.yaml]: Just so I'm not misunderstanding anything.  The header X-B3-Flags overrides the sampling policy on the service being instrumented (e.g. for debugging purposes to force a trace).  This is then recorded against the span for posterity.  There's no sampling policy per say on the zipkin service API itself correct?

Comment (adriancole) [inline zipkin-api.yaml]: Flags is a bit of an interesting one. In binary protocols, flags includes [more than the debug bit](https://github.com/openzipkin/brave/blob/master/brave-core/src/main/java/com/github/kristofa/brave/SpanId.java#L21). In http it ends up that X-B3-Flags is basically debug 0 or 1.

The constant `X-B3-Flags` is only defined in http header space, so it is accurate to say that debug is set on all spans in an http-originated trace that started with `X-B3-Flags: 1`.

sorry if this is too much info :)

Comment (singram) [inline zipkin-api.yaml]: Thanks for taking the time to explain.  Completely makes sense.  I was coming from a http centric world.  What was also interesting was the comment on parent/root spans where the parentId isn't null but equal to the current span and the traceId.  Seems to be many ways to skin the same cat which makes sense given the number of tracing libraries starting to use zipkin for storage and analysis.

Comment (adriancole) [inline zipkin-api.yaml]: yeah it is interesting :) I'm not sure the divergence between the binary
and http methods would be something folks would repeat, but they are what
they are. We can do it Right in B4 :P

Comment (singram) [inline zipkin-api.yaml]: Absolutely!  Let me know if there's any other changes to the branch as it stands I think I got all your comments.  Thanks again for spending the time on the review and conversation!

**Comment by adriancole** [comment]
nice job. I'll publish shortly

File: zipkin-api.yaml (hunk 8 above)
Comment (adriancole) [inline zipkin-api.yaml]: published. thx!

**Pull request merged**